src/dmd/dmodule.d

```diff
@@ -1288,11 +1250,35 @@ extern (C++) final class Module : Package
         return false;
     }
 
-    bool isRoot()
+    /**
+    A root module is one that will be compiled to object code.  This is used to
+    determine ownership of template instantiation and when to compile unittests.
+    Returns:
+        true if this is a root module
+    */
+    final bool isRoot() const
     {
         return this.importedFrom == this;
     }
 
+    /**
+    Called by the parser after the module declaration has been determined.
+    */
+    final void afterModuleDeclaration(ModuleDeclaration* md)
+    {
+        this.md = md;
+
+        // Call onImport here because if the module is going to be compiled then we
+        // need to determine it early because it affects whether unittests will be parsed
+        // and semantic analysis. This is being done right after parsing the module delclaration
+        // so the full module name can be taken from whatever was declared in the file.
+        if (!isRoot() && onImport && onImport(this))
+        {
+            importedFrom = this;
+            assert(isRoot());
+        }
```

Comment:
You could move `Target.loadModule` here also.

Reply:
Does the module need to be fully parsed before calling that, or does it just need the module declaration parsed, or does it not need anything to be parsed?

Reply:
Wait...I see from the comment it says "post parsing".  Which means it should be called in this function since this function is calling "during parsing" right after parsing the module declaration.

Reply:
Might be good to rename this function to something like `onModuleParsed` in another PR?